File: bench/passage_census.py
"""Counting pass of the §5.2.9 passage census (ticket 0140).

Consumes the JSONL that bench/passage_census.mjs measured with the embedder's
own tokenizer, applies the ratified geometry (bench/geometry.py — the one
implementation the fast tests cover), and writes the summary artifact §5.2.9
quotes: passages in total, per attachment kind, and the median per attachment
that derives band cap K.

Stated approximation: entries come from seg/1, which does not exist yet
(ticket 0028 builds to this geometry), so the census chunks each cache file as
one continuous paragraph sequence. Synthetic entry boundaries would only add
chunk closures, so the census errs low by that margin and the artifact says so.

    python3 bench/passage_census.py --paragraphs <paragraphs.jsonl> \
        --output bench/results/0140-passage-census/census.json
"""
import argparse
import json
import logging
import statistics
from pathlib import Path

import geometry

CENSUS_WINDOWS = Path(__file__).parent / "results" / "0140-model-windows" / "candidate-windows.json"

KINDS = ("pdf", "html", "other")


def summarize(rows: list[dict], budget: int) -> dict:
    per_kind = {k: {"files": 0, "paragraphs": 0, "tokens": 0, "passages": 0} for k in KINDS}
    per_file_passages = {k: [] for k in KINDS}
    for row in rows:
        kind = row["kind"]
        paragraphs = row["paragraphs"]
        passages = geometry.chunk_count(paragraphs, budget)
        bucket = per_kind[kind]
        bucket["files"] += 1
        bucket["paragraphs"] += len(paragraphs)
        bucket["tokens"] += sum(paragraphs)
        bucket["passages"] += passages
        per_file_passages[kind].append(passages)
    for kind in KINDS:
        counts = per_file_passages[kind]
        per_kind[kind]["median_passages_per_attachment"] = (
            statistics.median(counts) if counts else None
        )
    every = [c for kind in KINDS for c in per_file_passages[kind]]
    return {
        "budget": budget,
        "files": len(rows),
        "passages_total": sum(per_kind[k]["passages"] for k in KINDS),
        "tokens_total": sum(per_kind[k]["tokens"] for k in KINDS),
        "median_passages_per_attachment": statistics.median(every) if every else None,
        "by_kind": per_kind,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--paragraphs", type=Path, required=True,
                        help="JSONL from bench/passage_census.mjs")
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--windows", type=Path, default=CENSUS_WINDOWS,
                        help="window census the budget is resolved against")
    args = parser.parse_args()
    logging.basicConfig(level=logging.INFO, format="%(message)s")

    lines = args.paragraphs.read_text().splitlines()
    header = json.loads(lines[0])
    assert header.get("meta"), "first JSONL row must be the tokenization pass's meta header"
    rows = [json.loads(line) for line in lines[1:]]

    window = json.loads(args.windows.read_text())["min_window"]
    # The incumbent's passage prefix is empty (bench/models.json); a model with
    # an instruction prefix re-runs the tokenization pass, which records it.
    budget = geometry.resolve_budget(window, header["special_tokens"], 0)

    artifact = {
        "ticket": "tickets/0140-cap-the-chunker-below-the-embedder-limit.erg",
        "probe": ["bench/passage_census.mjs", "bench/passage_census.py"],
        "tokenization": {k: header[k] for k in
                         ("run_utc", "model", "special_tokens", "population", "sampled", "seed")},
        "geometry": {"budget": budget, "window": window,
                     "min_tokens": geometry.MIN_TOKENS, "overlap_tokens": geometry.OVERLAP_TOKENS},
        "approximation": "entry = one cache file (seg/1 not built yet); synthetic entry "
                         "boundaries would only add closures, so passage counts err low",
        "summary": summarize(rows, budget),
    }
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(artifact, indent=2) + "\n")
    logging.info(json.dumps(artifact["summary"], indent=2))


if __name__ == "__main__":
    main()

File: bench/passage_census.mjs
// Tokenization pass of the §5.2.9 passage census (ticket 0140): paragraph token
// counts for every fulltext cache, measured with the embedder's own tokenizer.
//
// Measure, do not divide: the ~1,35 tokens-per-word figure in the ticket was a
// labeled approximation, and the max rarely binds under structural chunking,
// so arithmetic on a ratio overstates any correction. This pass writes one
// JSONL row per cache file — attachment kind and the token count of each
// paragraph — and bench/passage_census.py turns those into passage counts with
// the geometry stated in bench/geometry.py. The split keeps the chunk-counting
// rule in one implementation, the tested Python one, instead of a JS twin
// that could drift.
//
// Tokenizer only — no ONNX session, no embedding: this pass counts, it does
// not time or embed, so its numbers cannot be corrupted by CPU contention
// (the run can still slow a concurrent TIMED measurement, which is a
// scheduling constraint, not a validity one).
//
//   node bench/passage_census.mjs --pkg-root <dir> \
//     --storage-root /path/to/Zotero/storage --output <paragraphs.jsonl> \
//     [--model all-minilm-l6-v2] [--sample 200] [--seed 0]
import { createRequire } from 'node:module';
import { appendFileSync, existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { pathToFileURL } from 'node:url';
import { parseArgs } from 'node:util';
import { resolveModel } from './registry.mjs';

const { values: opt } = parseArgs({
  options: {
    'pkg-root': { type: 'string' },
    'storage-root': { type: 'string' },
    output: { type: 'string' },
    model: { type: 'string', default: 'all-minilm-l6-v2' },
    // 0 = the whole corpus. A sample is for validating the pass, and its
    // artifact says so; §5.2.9 quotes only a full run.
    sample: { type: 'string', default: '0' },
    seed: { type: 'string', default: '0' },
  },
});
if (!opt['pkg-root'] || !opt['storage-root'] || !opt.output) {
  console.error(
    'usage: node bench/passage_census.mjs --pkg-root <dir> --storage-root <Zotero/storage> ' +
      '--output <paragraphs.jsonl> [--model M] [--sample N] [--seed S]',
  );
  process.exit(2);
}

const require = createRequire(`${opt['pkg-root'].replace(/\/?$/, '/')}package.json`);
const tjs = await import(pathToFileURL(require.resolve('@huggingface/transformers')).href);

const { id: modelId, repo: modelRepo } = resolveModel(opt.model);
const tok = await tjs.AutoTokenizer.from_pretrained(modelRepo);
const ntok = (s) => tok(s).input_ids.dims.at(-1);
const specialTokens = ntok('');

// Same attachment-kind rule as verification/probes/paragraph-size-distribution.py,
// because a third of the store is HTML snapshots whose extraction is chrome.
function kindOf(dir) {
  const exts = new Set(
    readdirSync(dir, { withFileTypes: true })
      .filter((e) => e.isFile() && !e.name.startsWith('.'))
      .map((e) => e.name.replace(/^.*(\.[^.]*)$/, '$1').toLowerCase()),
  );
  if (exts.has('.pdf')) return 'pdf';
  if (exts.has('.html') || exts.has('.htm') || exts.has('.mht')) return 'html';
  return 'other';
}

// Mulberry32: a seeded shuffle so a sample is reproducible.
function rng(seed) {
  let a = seed >>> 0;
  return () => {
    a |= 0; a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

let caches = readdirSync(opt['storage-root'], { withFileTypes: true })
  .filter((e) => e.isDirectory())
  .map((e) => join(opt['storage-root'], e.name, '.zotero-ft-cache'))
  .filter((p) => existsSync(p));

const population = caches.length;
const sampleSize = Number(opt.sample);
if (sampleSize > 0 && sampleSize < caches.length) {
  const random = rng(Number(opt.seed));
  for (let i = caches.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1));
    [caches[i], caches[j]] = [caches[j], caches[i]];
  }
  caches = caches.slice(0, sampleSize);
}

mkdirSync(dirname(opt.output), { recursive: true });
const header = {
  meta: true,
  probe: 'bench/passage_census.mjs',
  run_utc: new Date().toISOString(),
  model: { id: modelId, repo: modelRepo },
  special_tokens: specialTokens,
  storage_root: opt['storage-root'],
  population,
  sampled: caches.length,
  seed: Number(opt.seed),
};
writeFileSync(opt.output, `${JSON.stringify(header)}\n`);

let done = 0;
for (const cache of caches) {
  let text;
  try {
    text = readFileSync(cache, 'utf8');
  } catch {
    continue;
  }
  const key = dirname(cache).split('/').at(-1);
  const paragraphs = text
    .split(/\n\s*\n/)
    .map((p) => p.trim())
    .filter(Boolean)
    .map((p) => ntok(p) - specialTokens);
  appendFileSync(
    opt.output,
    `${JSON.stringify({ key, kind: kindOf(dirname(cache)), paragraphs })}\n`,
  );
  done += 1;
  if (done % 500 === 0) console.error(`${done}/${caches.length}`);
}
console.log(JSON.stringify({ population, processed: done, output: opt.output }));
